### PR TITLE
Vertically stretched grid (finally)

### LINKED ISSUE #471: Strategy for implementing and testing models with vertically stretched grids
I've figured out how to do the 3D pressure solve on a stretched grid following Chris' notes so I think we have all the pieces we need to implement vertically stretched grids, we just have to put it all together.

I'm thinking about how to go about fully incorporating a vertically stretched grid, true finite volume operators, and the new FFT+tridiagonal Poisson solvers needed to solve for the pressure on a stretched grid.

Seems like a good idea to split it up into steps with one pull request per step:
1. Reverse the `k` index. Currently PR #462.
2. Revise the `RegularCartesianGrid` struct so we're happy with it. Currently PR #464.
3. Add finite volume operators as a separate piece of code. Technically they won't be tested in this PR and could have mistakes. Currently PR #283
4. Nuke the old operators and start using the same set of finite volume operators for both `Oceananigans.Operators` and `closure_operators.jl`. This will test that the finite volume operators reduce down to the operators that currently work, but doesn't test them on a stretched grid.
5. Implement a `VerticallyStretchedCartesianGrid`. Might have to iterate bit to figure out what we need, e.g. I think we'll want `ΔzC` to include the distance between the first cell center and the halo cell center, etc.
6. Implement CPU and GPU pressure solvers for vertically stretched grids with tests. There will be two: one for horizontally periodic domains and another for channel models. I've figured most of this stuff out in Jupyter notebooks.
7. Ensure that models with vertically stretched grids pass basic tests: e.g. incompressibility, tracer conservation, etc. This will test the finite volume operators.
8. Run a model with a `VerticallyStretchedCartesianGrid` but with uniform grid spacing and make sure it produces the same numbers as a model with `RegularCartesianGrid`. This is a sanity check.
9. Run additional tests for vertically stretched grids: e.g. vertical diffusion, internal wave, etc. This will also test boundary conditions with stretched grids.
10. Rerun the stratified Couette flow verification experiment but with a stretching factor matching Vreugdenhil & Taylor (2018). This will test the AMD closure on stretched grids.

Let me know if anyone has any thoughts. cc @jm-c @rafferrari

### PR DESCRIPTION
**Pull request by ali-ramadhan**
This PR _finally_ revives the half-baked `VerticallyStretchedCartesianGrid` implementation that's been living in Oceananigans.jl for over a year.

It also formalizes the `FourierTridiagonalPoissonSolver` that's been living in `test_solvers.jl` for a while as well.

I re-ran the regression tests using a `VerticallyStretchedCartesianGrid` with constant spacing as a sanity check and they pass on the CPU which is good and helped uncover some functions that needed generalization.

There's still quite a bit to do and test so maybe we should merge this PR after a bit of polish and continue developing and testing the `VerticallyStretchedCartesianGrid` in subsequent smaller PRs?

TODO off the top of my head:
1. Adapt `VerticallyStretchedCartesianGrid` to work in GPU kernels. It has some offset arrays.
2. Test the `FourierTridiagonalPoissonSolver` for other/all topologies.
3. Add some dynamics tests?
4. Validation experiment: Stratified Couette flow (Vreugdenhil & Taylor, 2018).
5. Benchmarks! The xy FFT + z Tridiagonal solver might be faster than xyz FFT.

Resolves #471

And of course, no PR is complete without eye candy (had to use matplotlib for irregular heatmaps):

https://user-images.githubusercontent.com/20099589/107367287-1fc7b400-6aad-11eb-945f-1fb2694392b2.mp4

### CODE REVIEW

**Comment by francispoulin** [comment]
Beautiful eye candy!  Great job @ali-ramadhan and so looking forward to using this huge advancement.

**Comment by francispoulin** [comment]
A question @ali-ramadhan , how much work will it be to have a stretched grid in `x`?  I ask because I want to do a `ShallowWaterModel` run wtih a western boundary current and an intensified western grid.  I'm happy to work on this sometime but curious as to how much effort you think this will take, after it is fully tested of course.

**Comment by glwagner** [comment]
I think this PR should be re-opened right?

Does `VerticallyStretchedCartesianGrid` need an `arch` argument, or do we convert it within the constructor for `IncompressibleModel`?

File: src/Grids/vertically_stretched_cartesian_grid.jl [hunk 1]
```diff
@@ -90,13 +90,21 @@ function VerticallyStretchedCartesianGrid(FT=Float64, arch=CPU();
...
      zF = OffsetArray(zF,  -Hz)
     ΔzF = OffsetArray(ΔzF, -Hz)
 
+    # Needed for pressure solver solution to be divergence-free.
+    # Will figure out why later...
+    ΔzC[Nz] = ΔzC[Nz-1]
+
+    # Seems needed to avoid out-of-bounds error in viscous dissipation
+    # operator wanting to access ΔzC[Nz+2].
+    ΔzC = OffsetArray(cat(ΔzC[0], ΔzC..., ΔzC[Nz], dims=1), -Hz-1)
```
Comment (glwagner) [inline src/Grids/vertically_stretched_cartesian_grid.jl:106]: Trying to figure out if this makes sense...

File: src/Grids/vertically_stretched_cartesian_grid.jl [hunk 2]
```diff
@@ -149,3 +157,22 @@ function generate_stretched_vertical_grid(FT, z_topo, Nz, Hz, zF_generator)
 
     return Lz, zF, zC, ΔzF, ΔzC
 end
+
+# We cannot reconstruct a VerticallyStretchedCartesianGrid without the zF_generator.
+# So the best we can do is tell the user what they should have done.
```
Comment (glwagner) [inline src/Grids/vertically_stretched_cartesian_grid.jl:162]: Should we ask users to provide only the interior grid, and pad the interior grid with halo points according to the size of `halo`?

File: src/Operators/laplacian_operators.jl [hunk 3]
```diff
@@ -35,9 +35,9 @@ Calculates the Laplacian of c via
 which will end up at the location `ccc`.
 """
 @inline function ∇²(i, j, k, grid, c)
-    return 1/Vᵃᵃᶜ(i, j, k, grid) * (δxᶜᵃᵃ(i, j, k, grid, Ax_∂xᶠᵃᵃ, c) +
-                                    δyᵃᶜᵃ(i, j, k, grid, Ay_∂yᵃᶠᵃ, c) +
-                                    δzᵃᵃᶜ(i, j, k, grid, Az_∂zᵃᵃᶠ, c))
+    return 1/Vᵃᵃᶠ(i, j, k, grid) * (δxᶠᵃᵃ(i, j, k, grid, Ax_∂xᶜᵃᵃ, c) +
+                                    δyᵃᶠᵃ(i, j, k, grid, Ay_∂yᵃᶜᵃ, c) +
+                                    δzᵃᵃᶠ(i, j, k, grid, Az_∂zᵃᵃᶜ, c))
```
Comment (glwagner) [inline src/Operators/laplacian_operators.jl:40]: This only works for `Field{Face, Face, Face}`, right?

**Review by glwagner — APPROVED** [review]

File: src/Grids/vertically_stretched_cartesian_grid.jl (hunk 2 above)
Comment (francispoulin) [reply src/Grids/vertically_stretched_cartesian_grid.jl:162]: I'm a bit puzzeld by this change.  What is `zF_generator`?

If halo is set by what we need for the advection scheme, then I think we would need the interor halos to be as wide as the exterior halos.  Unless there is something that I'm missing?

File: src/Grids/vertically_stretched_cartesian_grid.jl (hunk 1 above)
Comment (ali-ramadhan) [reply src/Grids/vertically_stretched_cartesian_grid.jl:106]: Yeah I'm not sure either but everything passed when I added it so I just need to understand why it's needed. A similar issue cropped up for the `CompressibleModel` where we were sure we just needed halos of size 1 but the viscous dissipation operators wanted halos of at least size 2 because ρu and ρu / ρ create more interpolation.

File: src/Grids/vertically_stretched_cartesian_grid.jl (hunk 2 above)
Comment (ali-ramadhan) [reply src/Grids/vertically_stretched_cartesian_grid.jl:162]: Ah the `zF_generator` is either an array or function of `k` passed to `VerticallyStretchedCartesianGrid` that determines the locations of the faces in the z-direction.

The problem is that `with_halo(new_halo, old_grid)` is called in the model constructor after the grid has been created so the grid cannot be reconstructed via simple call to `VerticallyStretchedCartesianGrid` as the `zF_generator` is not stored in the grid.

We could probably get around this but I just decided to print a warning for now as I wanted to focus on getting tests to pass.

File: src/Operators/laplacian_operators.jl (hunk 3 above)
Comment (ali-ramadhan) [reply src/Operators/laplacian_operators.jl:40]: Or rather `Field{Center, Center, Center}` since it's used to compute ∇²ϕ.

I'm actually still a little confused which convention we're following. More comments and documentation is needed as you pointed out a long time ago: https://github.com/CliMA/Oceananigans.jl/issues/696

**Comment by ali-ramadhan** [comment]
> A question @ali-ramadhan , how much work will it be to have a stretched grid in `x`?  I ask because I want to do a `ShallowWaterModel` run wtih a western boundary current and an intensified western grid.  I'm happy to work on this sometime but curious as to how much effort you think this will take, after it is fully tested of course.

The hardest part of implementing a vertically-stretched grid for an incompressible model was implementing and testing the pressure solver. So it should be pretty straightforward to stretch x and y for `ShallowWaterModel`, just maybe a little tedious. 

Right now in `Oceananigans.Operators` we have functions `Δx(i, j, k, grid)` and `Δy(i, j, k, grid)` that return the grid spacings at grid point `[i, j, k]`.

Stretching the staggered grid in x and y we will need to add `ΔxC` and `ΔxF` functions (same for y) for the grid spacing between cell centers and cell faces and use them in the appropriate places (same for area and volume functions). This will probably mirror what has been for z with `ΔzF` and `ΔzC`.

Thankfully it'll probably be hard to break the code since there are so many tests in place that will fail if you miss a change.

File: src/Operators/laplacian_operators.jl (hunk 3 above)
Comment (glwagner) [reply src/Operators/laplacian_operators.jl:40]: I believe the 3-symbol code refers to the location of the _result_. So as written, this calculates the Laplacian of a function located at `Face, Face, Face` (which also ends up at the location `Face, Face, Face`).

File: src/Grids/vertically_stretched_cartesian_grid.jl (hunk 1 above)
Comment (glwagner) [reply src/Grids/vertically_stretched_cartesian_grid.jl:106]: Do `closure=nothing` tests pass with ordinary halos?

Could also be an off-by-one mistake somewhere ?

File: src/Grids/vertically_stretched_cartesian_grid.jl (hunk 2 above)
Comment (glwagner) [reply src/Grids/vertically_stretched_cartesian_grid.jl:162]: Ahh, I understand. Do we allow users to pass `zF` as an Array? In that case it might work to pass `grid.zF[1:Nz+1]` to the constructor for `VerticallyStretchedCartesianGrid`.

**Comment by ali-ramadhan** [comment]
If tests pass I'll merge this so it can be used with the hydrostatic model. Also don't want it to go stale again with operators being changed.

I'll open a subsequent PR that makes the vertically stretched grid work on the GPU.
